### PR TITLE
[4.2] Fix library autoloading

### LINKED ISSUE #39347: Namespace for 3d-party libraries with library name like Vendor/Libraryname don't work
It is assumed that `<libraryname>` is the name of the folder and the name of the xml manifest file. So `<libraryname>` should be `'mylib'`, the name of the folder in libraries will be `'mylib'`, the name of the manifest file is `"mylib.xml "`. The `'mylib'` folder will contain the `'src'` folder, which we specify in the `path `attribute of the `<namespace>` element.

However, we have the ability to specify a `<libraryname>` of the form `"iAmALibraryVendor/MyLib"`. If `<libraryname>` is specified in such a way that the library will be installed in the `"libraries/iAmALibraryVendor/MyLib"` folder. In Joomla 3, we registered namespace using a system plugin. We could install different sets of independent libraries in our `"iAmALibraryVendor"` folder, varying from project to project. Everything was ok. In Joomla 4, the installation of a library with such a `<libraryname>` occurs correctly. If you register namespace in the old way using the system plugin, then everything is ok.
But, we have the `<namespace>` element and we cannot use it in this case, since we cannot specify a slash in the name for the xml manifest file name. But this is necessary in order for the manifest file to load when updating the namespaces map.
 
See `libraries/namespacemap.php` function `getNamespaces `
https://github.com/joomla/joomla-cms/blob/f35ce22bd82d8db48834f9ac7168592335450fe0/libraries/namespacemap.php#L179

### PR DESCRIPTION
**Pull request by SharkyKZ**
### Summary of Changes

Fixes https://github.com/joomla/joomla-cms/issues/39347.

This fixes some issues in namespace mapper regarding library extension autoloading. First, this changes the manifest lookup from `JPATH_LIBRARIES` to `JPATH_MANIFESTS` directory where library manifests are actually stored. Secondly, this adds support for the possibly undocumented "vendor" library structure.

### Testing Instructions

See https://github.com/joomla/joomla-cms/discussions/39342 and https://github.com/joomla/joomla-cms/issues/39347.

### Documentation Changes Required

Although this PR is a bug fix, library autoloading and "vendor" library structure features are probably not documented. So documentation should be added somewhere.

### CODE REVIEW

**Comment by sergeytolkachyov** [comment]
@SharkyKZ tested it successfully.

**Comment by joomdonation** [comment]
@sergeytolkachyov Could you please go to https://issues.joomla.org/tracker/joomla-cms/39348 , click on Test this button and report your test result there so that your test will be counted? We require two successful tests for each PR before it can be merged.

**Comment by sergeytolkachyov** [comment]
I have tested this item :white_check_mark: successfully on 326bfe2e37ce2282d6b75b67ef2604cca8a9c5a3<br /><br />Tested this successfully.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39348">issues.joomla.org/tracker/joomla-cms/39348</a>.</sub>

**Comment by joomdonation** [comment]
Thanks @sergeytolkachyov

File: libraries/namespacemap.php [hunk 1]
```diff
@@ -178,113 +178,138 @@ protected function writeNamespaceFile($elements)
...
-                // Normalize the namespace string
-                $namespace     = str_replace('\\', '\\\\', $namespace) . '\\\\';
-                $namespacePath = rtrim($extensionPath . $namespaceNode->attributes()->path, '/');
+            return $manifests;
+        }
 
-                if ($type === 'plugin' || $type === 'library') {
-                    $baseDir = $type === 'plugin' ? 'JPATH_PLUGINS . \'' : 'JPATH_LIBRARIES . \'';
-                    $path    = str_replace($type === 'plugin' ? JPATH_PLUGINS : JPATH_LIBRARIES, '', $namespacePath);
+        $manifests = [];
```
Comment (richard67) [inline libraries/namespacemap.php]: Don't we have that already above in line 255?

Comment (SharkyKZ) [reply libraries/namespacemap.php]: Yes.

**Comment by richard67** [comment]
> I'll test it again tomorrow.

@sergeytolkachyov No need to do that. Your test is still valid, I restored the counter.

**Comment by sergeytolkachyov** [comment]
@richard67 what is currently missing for the merging of this pr?

**Comment by richard67** [comment]
> @richard67 what is currently missing for the merging of this pr?

@sergeytolkachyov A 2nd human test. Your test still counts, but each PR needs 2 human tests before it is set to RTC (ready to commit). After that a maintainer will merge it whenever they find time.

**Comment by sergeytolkachyov** [comment]
I have tested this item :white_check_mark: successfully on 53f4b886b7fc8386673773862f9c97869204a2df<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39348">issues.joomla.org/tracker/joomla-cms/39348</a>.</sub>

**Comment by leha123ra** [comment]
I have tested this item :white_check_mark: successfully on 53f4b886b7fc8386673773862f9c97869204a2df<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39348">issues.joomla.org/tracker/joomla-cms/39348</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39348">issues.joomla.org/tracker/joomla-cms/39348</a>.</sub>

**Comment by roland-d** [comment]
Thank you

**Pull request merged**